File: reference/patternPaint.ts
// patternPaint.ts — turn the built-in blob47 pattern into pixels.
//
// The caller picks ONE colour per role (terrain A, terrain B, outline); the
// shaded levels come from the recipes baked into blob47Pattern.ts, so there is
// nothing to tune and no field to evaluate — the silhouette is art data and
// this module only colours it in.
//
// Kept free of ImageData so it runs headless under vitest; App wraps the RGBA
// buffer when it needs to blit.

import {
  PATTERN_LEVELS,
  PATTERN_TILE_SIZE,
  SHADE_RECIPES,
  DEFAULT_BAND_STEPS,
  outlineWidthPx,
  bandNoiseSpan,
  patternLevelsFor,
  patternLevelsForMask,
  patternBandCoords,
  type PatternId,
  type PatternRole,
} from './blob47Pattern';
import {
  DEFAULT_NOISES, DEFAULT_NOISE_SEED, DEFAULT_NOISE_STRENGTH, DEFAULT_NOISE_TARGETS,
  noiseStep, type NoiseId, type NoiseTargetId,
} from './patternNoise';
import {
  NO_RIBBON, ribbonShadeAt, type RibbonOptions,
} from './patternRibbon';
import {
  DEFAULT_TEXTURE, DEFAULT_TEXTURE_SHADES, DEFAULT_CELL_SCALE, DEFAULT_RIPPLE_SCALE,
  DEFAULT_GEO_SCALE,
  textureColour, textureRamp, textureShadeAt, type TextureId,
} from './patternTexture';

/**
 * Speckle applied inside the solid terrains. `amount` 0 disables per terrain.
 *
 * The algorithm is per terrain as well as the amount and colour: the two solid
 * regions are different materials, and the whole point of the geometric ones is
 * that paving under grass wants a different field from the grass itself.
 */
export interface TextureOptions {
  algoA: TextureId;
  algoB: TextureId;
  amountA: number;
  amountB: number;
  shadesA: number;
  shadesB: number;
  seedA: number;
  seedB: number;
  cellScaleA?: number;
  cellScaleB?: number;
  rippleScaleA?: number;
  rippleScaleB?: number;
  /** Motif size for the generated geometric pavings; see GEO_SCALES. */
  geoScaleA?: number;
  geoScaleB?: number;
  /**
   * What each terrain's texture fades toward. Independent of the terrain and
   * band colours on purpose — the speckle in hand-drawn pixel art is usually a
   * different material, not a lighter version of the ground. Omitted means
   * derive it from the terrain colour (see textureRamp).
   */
  colourA?: RGB;
  colourB?: RGB;
  /**
   * Per-step overrides of the A/B ramps above, sparse and indexed from the bare
   * terrain at 0. Length must match that terrain's shade count + 1 or it is ignored — a stale
   * array from a different step count would recolour the wrong steps, which is
   * the same guard the band ramp carries.
   */
  rampA?: readonly (RGB | undefined)[];
  rampB?: readonly (RGB | undefined)[];
}

export const NO_TEXTURE: TextureOptions = {
  algoA: DEFAULT_TEXTURE,
  algoB: DEFAULT_TEXTURE,
  amountA: 0,
  amountB: 0,
  shadesA: DEFAULT_TEXTURE_SHADES,
  shadesB: DEFAULT_TEXTURE_SHADES,
  seedA: 0,
  seedB: 0,
  cellScaleA: DEFAULT_CELL_SCALE,
  cellScaleB: DEFAULT_CELL_SCALE,
  rippleScaleA: DEFAULT_RIPPLE_SCALE,
  rippleScaleB: DEFAULT_RIPPLE_SCALE,
  geoScaleA: DEFAULT_GEO_SCALE,
  geoScaleB: DEFAULT_GEO_SCALE,
};

export interface RGB {
  r: number;
  g: number;
  b: number;
}

export type RoleColours = Record<PatternRole, RGB>;

/**
 * The palette the shade recipes were solved against. It is a TEST FIXTURE, not
 * a UI default: the locked sheet hashes exist to catch a silhouette changing,
 * so they have to be measured against a palette that never moves. Changing what
 * the app starts up with must not disturb them.
 */
export const REFERENCE_ROLE_COLOURS: RoleColours = {
  terrainA: { r: 248, g: 248, b: 248 },
  terrainB: { r: 176, g: 216, b: 72 },
  edge: { r: 175, g: 198, b: 255 },
};

/**
 * What the app opens with: water on grass, with a sand shoreline.
 *
 * Terrain A is the painted region, so painting makes ponds and lakes in a grass
 * field. Terrain A's blue is deliberately short of full saturation — the
 * terrainA shade recipe only ADDS saturation (see SHADE_RECIPES), so a base at
 * s=1 has nowhere to go and the band's inner steps would collapse into the
 * terrain colour, leaving just the outline.
 */
export const DEFAULT_ROLE_COLOURS: RoleColours = {
  terrainA: { r: 58, g: 127, b: 201 },  // #3a7fc9 water
  terrainB: { r: 93, g: 168, b: 50 },   // #5da832 grass
  edge: { r: 232, g: 213, b: 160 },     // #e8d5a0 sand
};

/**
 * What the texture pickers open on: the shift the old auto-derivation would
 * have produced for the default terrains. Derived rather than written out, so
 * the starting point cannot drift away from textureColour().
 */
export const DEFAULT_TEXTURE_COLOURS: { terrainA: RGB; terrainB: RGB } = {
  terrainA: textureColour(DEFAULT_ROLE_COLOURS.terrainA, 1),
  terrainB: textureColour(DEFAULT_ROLE_COLOURS.terrainB, 1),
};

const clamp255 = (v: number) => Math.max(0, Math.min(255, Math.round(v)));

export function parseHexColour(hex: string): RGB {
  const s = hex.replace('#', '');
  const full = s.length === 3 ? s.split('').map((c) => c + c).join('') : s;
  return {
    r: parseInt(full.slice(0, 2), 16) || 0,
    g: parseInt(full.slice(2, 4), 16) || 0,
    b: parseInt(full.slice(4, 6), 16) || 0,
  };
}

export function toHexColour({ r, g, b }: RGB): string {
  return '#' + [r, g, b].map((v) => clamp255(v).toString(16).padStart(2, '0')).join('');
}

// --- HSV, matching the reference solve exactly ------------------------------
function rgbToHsv({ r, g, b }: RGB): [number, number, number] {
  const R = r / 255, G = g / 255, B = b / 255;
  const mx = Math.max(R, G, B);
  const mn = Math.min(R, G, B);
  const range = mx - mn;
  if (range === 0) return [0, 0, mx];
  const rc = (mx - R) / range;
  const gc = (mx - G) / range;
  const bc = (mx - B) / range;
  let h: number;
  if (R === mx) h = bc - gc;
  else if (G === mx) h = 2 + rc - bc;
  else h = 4 + gc - rc;
  h = (h / 6) % 1;
  return [h < 0 ? h + 1 : h, range / mx, mx];
}

function hsvToRgb(h: number, s: number, v: number): RGB {
  if (s === 0) return { r: clamp255(v * 255), g: clamp255(v * 255), b: clamp255(v * 255) };
  const i = Math.floor(h * 6);
  const f = h * 6 - i;
  const p = v * (1 - s);
  const q = v * (1 - s * f);
  const t = v * (1 - s * (1 - f));
  const table: [number, number, number][] = [
    [v, t, p], [q, v, p], [p, v, t], [p, q, v], [t, p, v], [v, p, q],
  ];
  const [R, G, B] = table[((i % 6) + 6) % 6];
  return { r: clamp255(R * 255), g: clamp255(G * 255), b: clamp255(B * 255) };
}

/**
 * A role's shaded variant. An achromatic base has no hue to shift, so the
 * recipe's `hue` is taken as an absolute one — that is what lets a plain white
 * terrain still pick up the cool rim tint the pattern was drawn with.
 */
export function shadeColour(c: RGB, role: PatternRole, t = 1): RGB {
  if (t <= 0) return c;
  const recipe = SHADE_RECIPES[role];
  const [h0, s0, v0] = rgbToHsv(c);
  // An achromatic base has no hue to nudge, so it takes the recipe's outright
  // and lets the scaled saturation decide how much of it shows. A coloured base
  // keeps its own hue and is only nudged — rotating it by the grey tint's hue
  // would send a deep blue off to magenta.
  const h = s0 < 1e-6 ? recipe.greyHue : (h0 + recipe.hue * t + 1) % 1;
  return hsvToRgb(
    h,
    Math.max(0, Math.min(1, s0 + recipe.sat * t)),
    Math.max(0, Math.min(1, v0 * (1 + (recipe.val - 1) * t)))
  );
}

/** The level colours, indexed by the digits in the level grid. */
export function patternRamp(colours: RoleColours, bandSteps?: number): RGB[] {
  const levels = bandSteps === undefined ? PATTERN_LEVELS : patternLevelsFor(bandSteps);
  return levels.map(({ role, shade }) =>
    shade > 0 ? shadeColour(colours[role], role, shade) : colours[role]
  );
}

/**
 * Everything that changes a tile's pixels beyond its pattern, mask and palette.
 *
 * An options object rather than the positional list this used to be: the band
 * refactor took it past sixteen parameters, at which point a call site is a row
 * of bare `undefined`s and adding one in the middle silently shifts the rest.
 */
export interface PaintOptions {
  /** Must be a multiple of PATTERN_TILE_SIZE — see the seam note below. */
  tileSize?: number;
  offsetPx?: number;
  bandSteps?: number;
  hardEdgeB?: boolean;
  /** Re-roll of the baked silhouette; 0 means exactly as baked. Also the
   *  ribbon motif's phase, because one edge should answer to one dice. */
  edgeSeed?: number;
  outlineWidth?: number;
  /** Grain on the transition band. The algorithms stack; empty means none. */
  noises?: readonly NoiseId[];
  noiseSeed?: number;
  noiseStrength?: number;
  /** Which of the band's three zones the grain may move a pixel out of. */
  noiseTargets?: readonly NoiseTargetId[];
  /** Picked colours for the grain, one per direction it nudges a pixel. */
  noiseColours?: { b?: RGB; edge?: RGB; a?: RGB };
  /** What is painted inside the outline. */
  ribbon?: RibbonOptions;
  /** Speckle inside the two solid terrains. */
  texture?: TextureOptions;
  /** Overrides the whole derived level ramp; ignored unless it matches length. */
  ramp?: readonly RGB[];
  /**
   * Paint nothing where terrain B would go, so the sheet can be stacked over
   * another tile layer.
   *
   * It is the ROLE that goes transparent, not the level: terrain B owns two
   * levels — the open field and the one-pixel shaded rim hugging the outline —
   * and both vanish. Keeping the rim would mean painting a shaded variant of a
   * colour that is not being drawn, i.e. a coloured halo round every tile with
   * nothing to explain it, and it would put opaque pixels outside the outline
   * where a layer beneath is supposed to show through.
   *
   * Grain follows the same rule, decided by where a pixel LANDS rather than
   * where it came from: an outline pixel the grain pushes out into terrain B
   * becomes a hole, which is what "dissolving the edge outward" has to mean once
   * there is nothing behind it. That also makes the picked terrain-B grain
   * colour inert, and the panel greys it out.
   */
  transparentB?: boolean;
}

/**
 * One tile as RGBA bytes.
 *
 * `tileSize` must be a multiple of PATTERN_TILE_SIZE. That is not cosmetic: the
 * texture fields repeat every 16 or 32 output pixels, so seams (which fall every
 * `tileSize` pixels) only line up with them when that period divides `tileSize`.
 */
export function paintPatternTileRGBA(
  pattern: PatternId,
  mask: number,
  colours: RoleColours,
  opts: PaintOptions = {}
): Uint8ClampedArray<ArrayBuffer> {
  const {
    tileSize = PATTERN_TILE_SIZE,
    offsetPx = 0,
    bandSteps = DEFAULT_BAND_STEPS,
    hardEdgeB = false,
    edgeSeed = 0,
    outlineWidth,
    noises = DEFAULT_NOISES,
    noiseSeed = DEFAULT_NOISE_SEED,
    noiseStrength = DEFAULT_NOISE_STRENGTH,
    noiseTargets = DEFAULT_NOISE_TARGETS,
    noiseColours,
    ribbon = NO_RIBBON,
    texture = NO_TEXTURE,
    ramp: customRamp,
    transparentB = false,
  } = opts;

  const derived = patternRamp(colours, bandSteps);
  const ramp = customRamp && customRamp.length === derived.length ? customRamp : derived;
  const levelDefs = patternLevelsFor(bandSteps);
  const grid = patternLevelsForMask(
    pattern, mask, offsetPx, tileSize, bandSteps, hardEdgeB, edgeSeed, outlineWidth
  );
  const solid = ramp.length - 1;

  // Texture shades are a handful of colours, not a per-pixel computation.
  const fitRamp = (r: readonly (RGB | undefined)[] | undefined, n: number) =>
    r && r.length === n + 1 ? r : undefined;
  const shadesA = Math.max(1, texture.shadesA);
  const shadesB = Math.max(1, texture.shadesB);
  const texA = texture.algoA !== 'none' && texture.amountA > 0
    ? textureRamp(colours.terrainA, texture.colourA, shadesA, fitRamp(texture.rampA, shadesA))
    : null;
  // Not built at all when terrain B is transparent — there is no surface to
  // texture, and every pixel it would touch is about to be discarded anyway.
  const texB = !transparentB && texture.algoB !== 'none' && texture.amountB > 0
    ? textureRamp(colours.terrainB, texture.colourB, shadesB, fitRamp(texture.rampB, shadesB))
    : null;

  // The outline is one level, whatever the step count — extra steps are added
  // on the terrain-A side, so its index is found rather than assumed.
  const edgeLevel = levelDefs.findIndex((l) => l.role === 'edge');
  const ribShades = Math.max(1, ribbon.shades);
  const ribbonOn = ribbon.algo !== 'none' && ribbon.amount > 0 && mask >= 0;
  const ribRamp = ribbonOn
    ? textureRamp(ramp[edgeLevel], ribbon.colour, ribShades, fitRamp(ribbon.ramp, ribShades))
    : null;
  const coords = ribbonOn
    ? patternBandCoords(
        pattern, mask, offsetPx, tileSize, bandSteps, hardEdgeB, edgeSeed, outlineWidth
      )
    : null;
  const ribWidth = ribbonOn
    ? Math.max(1, outlineWidthPx(pattern, bandSteps, hardEdgeB, outlineWidth, tileSize))
    : 1;

  // Grain displacement scales with the band so it keeps reading as the band
  // widens; at the default step count this is 1 and nothing changes.
  const span = bandNoiseSpan(pattern, bandSteps);

  // Backed by a plain ArrayBuffer (not ArrayBufferLike) so the result can be
  // handed straight to `new ImageData(...)`.
  const out = new Uint8ClampedArray(new ArrayBuffer(tileSize * tileSize * 4));
  for (let y = 0; y < tileSize; y++) {
    for (let x = 0; x < tileSize; x++) {
      const p = y * tileSize + x;
      const level = grid.charCodeAt(p) - 48;
      let rgb = ramp[level];
      // Where the pixel ends up, which is what decides its alpha. Only grain
      // moves it: the ribbon and the textures repaint within a level.
      let finalLevel = level;
      // Grain lives on the transition band only, and is sampled in OUTPUT
      // space so it gets finer along with the art instead of blocking up.
      let grained = false;
      if (level > 0 && level < solid && noises.length > 0) {
        const step = noiseStep(noises, x, y, noiseSeed, noiseStrength) * span;

        if (step !== 0) {
          const nextLvl = Math.max(0, Math.min(solid, level + step));
          const fromRole = levelDefs[level]?.role;
          const nextRole = levelDefs[nextLvl]?.role;

          // The band is partitioned into three zones — the outline, the
          // terrain-A side and the terrain-B side — and a displacement touches
          // TWO of them, so both ends are checked. Either one alone leaks, and
          // both leaks are visible on the outline:
          //
          //   source only      a B-side pixel is promoted INTO the outline
          //                    level, scattering outline-coloured specks along
          //                    the band;
          //   destination only an outline pixel is demoted to the B side,
          //                    punching terrain-coloured holes in the outline.
          //
          // The guard is deliberately asymmetric rather than "both roles must
          // be targeted": moving OUT of a targeted zone is always allowed, so
          // that targeting the outline alone still dissolves it outward instead
          // of becoming a checkbox that does nothing. Only moving INTO the
          // outline is gated, because the outline is the one zone whose job is
          // to stay a readable line.
          const keepNoise =
            Boolean(fromRole && noiseTargets.includes(fromRole)) &&
            (nextRole !== 'edge' || noiseTargets.includes('edge'));

          if (keepNoise) {
            grained = true;
            finalLevel = nextLvl;
            if (nextRole === 'edge' && noiseColours?.edge) {
              rgb = noiseColours.edge;
            } else if (step < 0 && noiseColours?.b) {
              rgb = noiseColours.b;
            } else if (step > 0 && noiseColours?.a) {
              rgb = noiseColours.a;
            } else {
              rgb = ramp[nextLvl];
            }
          }
        }
      }
      // A grained pixel has already left the outline, so the motif does not get
      // to paint over it — the grain is what is eating the line, and a motif
      // drawn on top would fill the holes back in.
      if (!grained && ribRamp && coords && level === edgeLevel) {
        const k = ribbonShadeAt(
          ribbon.algo, coords.s[p], coords.depth[p], ribWidth,
          ribbon.seed, ribbon.amount, ribShades, ribbon.period, ribbon.invert
        );
        if (k > 0) rgb = ribRamp[k];
      } else if (texA && level === solid) {
        const k = textureShadeAt(
          texture.algoA, x, y, texture.seedA, texture.amountA, shadesA,
          texture.cellScaleA ?? DEFAULT_CELL_SCALE, texture.rippleScaleA ?? DEFAULT_RIPPLE_SCALE,
          texture.geoScaleA ?? DEFAULT_GEO_SCALE
        );
        if (k > 0) rgb = texA[k];
      } else if (texB && level === 0) {
        const k = textureShadeAt(
          texture.algoB, x, y, texture.seedB, texture.amountB, shadesB,
          texture.cellScaleB ?? DEFAULT_CELL_SCALE, texture.rippleScaleB ?? DEFAULT_RIPPLE_SCALE,
          texture.geoScaleB ?? DEFAULT_GEO_SCALE
        );
        if (k > 0) rgb = texB[k];
      }
      const { r, g, b } = rgb;
      const i = p * 4;
      out[i] = r;
      out[i + 1] = g;
      out[i + 2] = b;
      // Hard alpha, never a blend. A tileset is stacked and re-sampled, and a
      // half-transparent pixel picks up whatever happened to be underneath at
      // export time; pixel art keeps its silhouette in the alpha channel exactly
      // the way it keeps its colours in the palette.
      out[i + 3] = transparentB && levelDefs[finalLevel]?.role === 'terrainB' ? 0 : 255;
    }
  }
  return out;
}

File: src/pattern/pattern_paint.cpp
#include "pattern_paint.h"
#include "pattern_texture.h"
#include "pattern_ribbon.h"
#include "catalog.h"
#include "js_math.h"
#include <algorithm>
#include <cmath>
#include <string_view>

namespace atm {

const RoleColours DEFAULT_ROLE_COLOURS = {
    {  58, 127, 201 }, // #3a7fc9 water
    {  93, 168,  50 }, // #5da832 grass
    { 232, 213, 160 }  // #e8d5a0 sand
};

static inline int nibble(char c) {
    if (c >= '0' && c <= '9') return c - '0';
    if (c >= 'a' && c <= 'f') return c - 'a' + 10;
    if (c >= 'A' && c <= 'F') return c - 'A' + 10;
    return 0;
}

RGB parse_hex_colour(const std::string& hex) {
    std::string_view s = hex;
    if (!s.empty() && s[0] == '#') s.remove_prefix(1);
    if (s.length() == 3) {
        int r = nibble(s[0]);
        int g = nibble(s[1]);
        int b = nibble(s[2]);
        return { static_cast<uint8_t>(r * 17), static_cast<uint8_t>(g * 17), static_cast<uint8_t>(b * 17) };
    }
    if (s.length() >= 6) {
        int r = (nibble(s[0]) << 4) | nibble(s[1]);
        int g = (nibble(s[2]) << 4) | nibble(s[3]);
        int b = (nibble(s[4]) << 4) | nibble(s[5]);
        return { static_cast<uint8_t>(r), static_cast<uint8_t>(g), static_cast<uint8_t>(b) };
    }
    return { 0, 0, 0 };
}

std::string to_hex_colour(RGB c) {
    static const char HEX_DIGITS[] = "0123456789abcdef";
    return { '#',
             HEX_DIGITS[(c.r >> 4) & 0xf], HEX_DIGITS[c.r & 0xf],
             HEX_DIGITS[(c.g >> 4) & 0xf], HEX_DIGITS[c.g & 0xf],
             HEX_DIGITS[(c.b >> 4) & 0xf], HEX_DIGITS[c.b & 0xf] };
}

std::tuple<double, double, double> rgb_to_hsv(RGB c) {
    double R = static_cast<double>(c.r) / 255.0;
    double G = static_cast<double>(c.g) / 255.0;
    double B = static_cast<double>(c.b) / 255.0;
    double mx = std::max({ R, G, B });
    double mn = std::min({ R, G, B });
    double range = mx - mn;
    if (range == 0.0) return { 0.0, 0.0, mx };
    double rc = (mx - R) / range;
    double gc = (mx - G) / range;
    double bc = (mx - B) / range;
    double h;
    if (R == mx) h = bc - gc;
    else if (G == mx) h = 2.0 + rc - bc;
    else h = 4.0 + gc - rc;
    h = std::fmod(h / 6.0, 1.0);
    return { h < 0.0 ? h + 1.0 : h, range / mx, mx };
}

RGB hsv_to_rgb(double h, double s, double v) {
    auto cl = [](double x) {
        return static_cast<uint8_t>(std::max(0.0, std::min(255.0, std::floor(x + 0.5))));
    };
    if (s == 0.0) {
        uint8_t val = cl(v * 255.0);
        return { val, val, val };
    }
    double i_f = std::floor(h * 6.0);
    int i = static_cast<int>(i_f);
    double f = h * 6.0 - i_f;
    double p = v * (1.0 - s);
    double q = v * (1.0 - s * f);
    double t = v * (1.0 - s * (1.0 - f));
    double table[6][3] = {
        { v, t, p }, { q, v, p }, { p, v, t }, { p, q, v }, { t, p, v }, { v, p, q }
    };
    int idx = ((i % 6) + 6) % 6;
    return { cl(table[idx][0] * 255.0), cl(table[idx][1] * 255.0), cl(table[idx][2] * 255.0) };
}

RGB shade_colour(RGB c, PatternRole role, float t_f) {
    if (t_f <= 0.0f) return c;
    double t = static_cast<double>(t_f);
    struct ShadeRecipe {
        double hue, greyHue, sat, val;
    };
    ShadeRecipe recipe;
    if (role == PatternRole::TerrainA) {
        recipe = { 0.0, 0.541667, 0.129032, 1.000000 };
    } else if (role == PatternRole::TerrainB) {
        recipe = { 0.012037, 0.012037, 0.166667, 0.888889 };
    } else {
        recipe = { 0.0, 0.0, 0.0, 1.0 };
    }

    auto [h0, s0, v0] = rgb_to_hsv(c);
    double h = (s0 < 1e-6) ? recipe.greyHue : std::fmod(h0 + recipe.hue * t + 1.0, 1.0);
    return hsv_to_rgb(
        h,
        std::max(0.0, std::min(1.0, s0 + recipe.sat * t)),
        std::max(0.0, std::min(1.0, v0 * (1.0 + (recipe.val - 1.0) * t)))
    );
}

std::vector<RGB> pattern_ramp(const RoleColours& colours, int band_steps) {
    auto levels = pattern_levels_for(band_steps);
    std::vector<RGB> out;
    out.reserve(levels.size());
    for (const auto& l : levels) {
        RGB base = (l.role == PatternRole::TerrainA) ? colours.terrainA
                 : (l.role == PatternRole::TerrainB) ? colours.terrainB
                 : colours.edge;
        if (l.shade > 0.0f) {
            out.push_back(shade_colour(base, l.role, l.shade));
        } else {
            out.push_back(base);
        }
    }
    return out;
}

struct GrainResult {
    RGB rgb;
    int final_level;
    bool grained;
};

/**
 * Procedural grain / halftone noise displacement per-pixel.
 *
 * Checks whether the current pixel level qualifies for noise displacement
 * according to `opts.noise_targets` (TerrainA, TerrainB, Edge). If perturbed,
 * steps the quantized level by `step * span` and applies either custom noise
 * color overrides (e.g. `opts.noise_colours.edge`) or samples the adjusted ramp.
 */
template <typename TargetMatcher>
static GrainResult apply_grain(
    int level,
    int x,
    int y,
    const PaintOptions& opts,
    const std::vector<RGB>& ramp,
    const std::vector<PatternLevelDef>& level_defs,
    int solid,
    int span,
    const TargetMatcher& target_matches
) {
    if (level <= 0 || level >= solid || opts.noises.empty()) {
        return { ramp[level], level, false };
    }

    int step = noise_step(opts.noises, x, y, opts.noise_seed, static_cast<float>(opts.noise_strength)) * span;
    if (step == 0) {
        return { ramp[level], level, false };
    }

    int next_lvl = std::max(0, std::min(solid, level + step));
    PatternRole from_role = level_defs[level].role;
    PatternRole next_role = level_defs[next_lvl].role;

    bool keep_noise = target_matches(from_role) &&
        (next_role != PatternRole::Edge || target_matches(PatternRole::Edge));

    if (!keep_noise) {
        return { ramp[level], level, false };
    }

    RGB rgb;
    if (next_role == PatternRole::Edge && opts.noise_colours.edge.has_value()) {
        rgb = *opts.noise_colours.edge;
    } else if (step < 0 && opts.noise_colours.b.has_value()) {
        rgb = *opts.noise_colours.b;
    } else if (step > 0 && opts.noise_colours.a.has_value()) {
        rgb = *opts.noise_colours.a;
    } else {
        rgb = ramp[next_lvl];
    }
    return { rgb, next_lvl, true };
}

/**
 * Selects surface texture or ribbon motif overlay color at the given pixel.
 * Priority hierarchy:
 *   1. Ribbon Motif (if on Edge level)
 *   2. Texture A (if on interior solid TerrainA level)
 *   3. Texture B (if on exterior background TerrainB level 0)
 */
std::optional<RGB> TilePainter::pick_overlay(
    int level,
    int x,
    int y,
    int p,
    const std::optional<BandCoords>& coords
) const {
    if (ribbon_ramp_.has_value() && coords.has_value() && level == edge_level_) {
        int k = ribbon_shade_at(
            opts_.ribbon.algo, coords->s[p], coords->depth[p], ribbon_width_,
            opts_.ribbon.seed, opts_.ribbon.amount, ribbon_shades_, opts_.ribbon.period, opts_.ribbon.invert
        );
        if (k > 0 && k < static_cast<int>(ribbon_ramp_->size())) {
            return (*ribbon_ramp_)[k];
        }
    } else if (texture_ramp_a_.has_value() && level == solid_) {
        int k = texture_shade_at(
            opts_.texture.a.algo, x, y, opts_.texture.a.seed, static_cast<float>(opts_.texture.a.amount),
            std::max(1, opts_.texture.a.shades),
            opts_.texture.a.cellScale, opts_.texture.a.rippleScale, opts_.texture.a.geoScale
        );
        if (k > 0 && k < static_cast<int>(texture_ramp_a_->size())) {
            return (*texture_ramp_a_)[k];
        }
    } else if (texture_ramp_b_.has_value() && level == 0) {
        int k = texture_shade_at(
            opts_.texture.b.algo, x, y, opts_.texture.b.seed, static_cast<float>(opts_.texture.b.amount),
            std::max(1, opts_.texture.b.shades),
            opts_.texture.b.cellScale, opts_.texture.b.rippleScale, opts_.texture.b.geoScale
        );
        if (k > 0 && k < static_cast<int>(texture_ramp_b_->size())) {
            return (*texture_ramp_b_)[k];
        }
    }
    return std::nullopt;
}

TilePainter::TilePainter(const std::string& pattern, const RoleColours& colours, const PaintOptions& opts)
    : pattern_(pattern), colours_(colours), opts_(opts)
{
    auto derived = pattern_ramp(colours_, opts_.band_steps);
    ramp_ = (opts_.ramp.has_value() && opts_.ramp->size() == derived.size()) ? *opts_.ramp : derived;
    level_defs_ = pattern_levels_for(opts_.band_steps);
    fp_ = FieldParams{
        static_cast<float>(opts_.offset_px),
        opts_.tile_size,
        opts_.band_steps,
        opts_.hard_edge_b,
        opts_.edge_seed,
        static_cast<float>(opts_.outline_width)
    };
    solid_ = static_cast<int>(ramp_.size()) - 1;

    if (opts_.texture.a.algo != "none" && opts_.texture.a.amount > 0.0) {
        texture_ramp_a_ = texture_ramp(colours_.terrainA, opts_.texture.a.colour, std::max(1, opts_.texture.a.shades), opts_.texture.a.ramp);
    }

    if (!opts_.transparent_b && opts_.texture.b.algo != "none" && opts_.texture.b.amount > 0.0) {
        texture_ramp_b_ = texture_ramp(colours_.terrainB, opts_.texture.b.colour, std::max(1, opts_.texture.b.shades), opts_.texture.b.ramp);
    }

    edge_level_ = -1;
    for (size_t l = 0; l < level_defs_.size(); ++l) {
        if (level_defs_[l].role == PatternRole::Edge) {
            edge_level_ = static_cast<int>(l);
            break;
        }
    }

    ribbon_shades_ = std::max(1, opts_.ribbon.shades);
    ribbon_enabled_ = (opts_.ribbon.algo != "none" && opts_.ribbon.amount > 0.0);

    if (ribbon_enabled_ && edge_level_ >= 0) {
        ribbon_ramp_ = texture_ramp(ramp_[edge_level_], opts_.ribbon.colour, ribbon_shades_, opts_.ribbon.ramp);
    }

    ribbon_width_ = ribbon_enabled_ ? std::max(1.0f, outline_width_px(pattern_, opts_.band_steps, opts_.hard_edge_b, static_cast<float>(opts_.outline_width), opts_.tile_size)) : 1.0f;
    noise_span_ = band_noise_span(pattern_, opts_.band_steps);

    static_assert(static_cast<int>(PatternRole::Edge) == 2,
                  "noise_targets_lut_ is indexed by PatternRole; keep it dense and sized to match");
    noise_targets_lut_.fill(false);
    for (auto tid : opts_.noise_targets) {
        if (tid == NoiseTargetId::TerrainA) noise_targets_lut_[static_cast<int>(PatternRole::TerrainA)] = true;
        if (tid == NoiseTargetId::TerrainB) noise_targets_lut_[static_cast<int>(PatternRole::TerrainB)] = true;
        if (tid == NoiseTargetId::Edge)     noise_targets_lut_[static_cast<int>(PatternRole::Edge)] = true;
    }
}

void TilePainter::paint_tile_into(int mask, uint8_t* out_rgba, int row_stride_bytes) const {
    int tile_size = opts_.tile_size;
    int stride = (row_stride_bytes > 0) ? row_stride_bytes : (tile_size * 4);

    std::string grid = pattern_levels_for_mask(pattern_, mask, fp_);

    std::optional<BandCoords> coords;
    if (ribbon_enabled_ && mask >= 0) {
        coords = pattern_band_coords(pattern_, mask, fp_);
    }

    auto target_matches = [this](PatternRole r) {
        return noise_targets_lut_[static_cast<size_t>(r)];
    };

    for (int y = 0; y < tile_size; ++y) {
        uint8_t* row_ptr = out_rgba + y * stride;
        for (int x = 0; x < tile_size; ++x) {
            int p = y * tile_size + x;
            int level = grid[p] - '0';

            auto grain = apply_grain(level, x, y, opts_, ramp_, level_defs_, solid_, noise_span_, target_matches);
            RGB rgb = grain.rgb;
            if (!grain.grained) {
                if (auto overlay = pick_overlay(level, x, y, p, coords)) {
                    rgb = *overlay;
                }
            }

            int i = x * 4;
            row_ptr[i]     = rgb.r;
            row_ptr[i + 1] = rgb.g;
            row_ptr[i + 2] = rgb.b;
            row_ptr[i + 3] = (opts_.transparent_b && level_defs_[grain.final_level].role == PatternRole::TerrainB) ? 0 : 255;
        }
    }
}

std::vector<uint8_t> TilePainter::paint_tile_rgba(int mask) const {
    std::vector<uint8_t> out(opts_.tile_size * opts_.tile_size * 4, 0);
    paint_tile_into(mask, out.data(), opts_.tile_size * 4);
    return out;
}

std::vector<uint8_t> paint_pattern_tile_rgba(
    const std::string& pattern,
    int mask,
    const RoleColours& colours,
    const PaintOptions& opts
) {
    return TilePainter(pattern, colours, opts).paint_tile_rgba(mask);
}

} // namespace atm
